agent/php_txn.c

```diff
@@ -1109,3 +1109,101 @@ nr_status_t nr_php_txn_end(int ignoretxn, int in_post_deactivate TSRMLS_DC) {
 
   return NR_SUCCESS;
 }
+
+extern void nr_php_txn_add_code_level_metrics(
+    nr_attributes_t* attributes,
+    const nr_php_execute_metadata_t* metadata) {
+/* Current CLM functionality only works with PHP 7+ */
+#if ZEND_MODULE_API_NO < ZEND_7_0_X_API_NO /* PHP < PHP7 */
+  (void)attributes;
+  (void)metadata;
+  return;
+}
+#else
+  /*
+   * Check if code level metrics are enabled in the ini.
+   * If they aren't, exit and don't add any attributes.
+   */
+  if (!NRINI(code_level_metrics_enabled)) {
+    return;
+  }
+
+  if (NULL == metadata) {
+    return;
+  }
+
+  /*
+   * At a minimum, at least one of the following attribute combinations MUST be
+   * implemented in order for customers to be able to accurately identify their
+   * instrumented functions:
+   *  - code.filepath AND code.function
+   *  - code.namespace AND code.function
+   *
+   * If we don't have the minimum requirements, exit and don't add any
+   * attributes.
+   *
+   * Additionally, none of the needed attributes can exceed 255 characters.
+   */
+
+#define CHK_CLM_STRLEN(s, zstr_len) \
+  if (CLM_STRLEN_MAX < zstr_len) {  \
+    s = NULL;                       \
+  }
+
+  const char* namespace = NULL;
+  const char* function = NULL;
+  const char* filepath = NULL;
+
+  if (NULL != metadata->scope) {
+    namespace = ZSTR_VAL(metadata->scope);
+    CHK_CLM_STRLEN(namespace, ZSTR_LEN(metadata->scope));
+  }
+
+  if (NULL != metadata->function) {
+    function = ZSTR_VAL(metadata->function);
+    CHK_CLM_STRLEN(function, ZSTR_LEN(metadata->function));
+  }
+
+  if (NULL != metadata->filepath) {
+    filepath = ZSTR_VAL(metadata->filepath);
+    CHK_CLM_STRLEN(filepath, ZSTR_LEN(metadata->filepath));
+  }
+
+#undef CHK_CLM_STRLEN
+
+  if (1 == metadata->function_lineno) {
+    /*
+     * It's a file.  For CLM purposes, the "function" name is the filepath.
+     */
+    function = filepath;
+  }
+
+  if (nr_strempty(function)) {
+    /*
+     * Name isn't set so don't do anything
+     */
+    return;
+  }
+  if (nr_strempty(namespace) && nr_strempty(filepath)) {
+    /*
+     * CLM MUST have either function+namespace or function+filepath.
+     */
+    return;
+  }
+
+  nr_txn_attributes_set_string_attribute(attributes, nr_txn_clm_code_function,
+                                         function);
+
+  if (!nr_strempty(filepath)) {
+    nr_txn_attributes_set_string_attribute(attributes, nr_txn_clm_code_filepath,
+                                           filepath);
+  }
+  if (!nr_strempty(namespace)) {
+    nr_txn_attributes_set_string_attribute(
+        attributes, nr_txn_clm_code_namespace, namespace);
+  }
+
+  nr_txn_attributes_set_long_attribute(attributes, nr_txn_clm_code_lineno,
+                                       metadata->function_lineno);
```

Comment:
This code could be simplified to this:
```
#define CLM_ATTRIBUTE_DESTINATION\
  (NR_ATTRIBUTE_DESTINATION_TXN_TRACE | NR_ATTRIBUTE_DESTINATION_ERROR \
   | NR_ATTRIBUTE_DESTINATION_TXN_EVENT | NR_ATTRIBUTE_DESTINATION_SPAN)

#define SEGMENT_ATTRIBUTES_ADD_CLM(clm_attribute_name, clm_attribute_value, type) do { \
  nr_attributes_agent_add_##type(attributes, \
      CLM_ATTRIBUTE_DESTINATION, \
      clm_attribute_name, \
      clm_attribute_value); \
} while (0);

  SEGMENT_ATTRIBUTES_ADD_CLM("code.function", function, string);
  SEGMENT_ATTRIBUTES_ADD_CLM("code.filepath", filepath, string);
  SEGMENT_ATTRIBUTES_ADD_CLM("code.namespace", namespace, string);
  SEGMENT_ATTRIBUTES_ADD_CLM("code.lineno", metadata->function_lineno, long);
```
This would simplify a lot of code and re-use existing and tested functions:
1. Keep all CLM related things (attribute names and attribute destinations) in a single place - the function that adds the attributes.
2. Re-use existing and already tested functions: `nr_attributes_agent_add_string` and `nr_attributes_agent_add_long`  (reduced test maintenance).
3. Remove [these lines](https://github.com/zsistla/newrelic-php-agent/blob/clm_to_dev/axiom/nr_txn.c#L108-L119), [these functions](https://github.com/zsistla/newrelic-php-agent/blob/clm_to_dev/axiom/nr_txn.c#L174-L198) and tests for them.

Reply:
Yes, now it's been decoupled from txn->attributes, we can simplify a bit more.

Reply:
Please see: [b05f098](https://github.com/newrelic/newrelic-php-agent/pull/585/commits/b05f0980b22385746503dc4272ca59d8228eccb8)

While we can localize destinations and attribute names and remove no longer required functions, in practice, the macro doesn't seem to save us much since it is a misdirection to the actual function call and we still have to check file path and namespace for strempty but we don't need to check function or lineno.

It would still need to look like this and in this case the macro seems to hide more than help because we are passing to a macro to just pass straight to a function:
```
SEGMENT_ATTRIBUTES_ADD_CLM("code.function", function, string);

/*
 * If the string is empty, CLM specs say don't add it.
 * nr_attributes_agent_add_string is okay with an empty string attribute.
 * Already checked function for strempty no need to check again.
 */
if (!nr_strempty(filepath)) {
  SEGMENT_ATTRIBUTES_ADD_CLM("code.filepath", filepath, string);
}

if (!nr_strempty(namespace)) {
  SEGMENT_ATTRIBUTES_ADD_CLM("code.namespace", namespace, string);
}

SEGMENT_ATTRIBUTES_ADD_CLM("code.lineno", metadata->function_lineno, long);
```

Seems in this case, better to just be direct: https://github.com/newrelic/newrelic-php-agent/pull/585/commits/b05f0980b22385746503dc4272ca59d8228eccb8#diff-beedad7bbfc9687348e557b4752fbf6f54e4407a6e3177e8cabaaec7e3ffd625R1169

Reply:
Fair enough, looks much cleaner now! Thank you!